### PR TITLE
Add MbedProgram and metadata classes.

### PR DESCRIPTION
**Pull request by rwalton-arm**
### Description

This commit adds a set of objects which represent elements in an mbed-project
folder.

MbedProgram: Models an Mbed program. Provides classmethods to handle
various initialisation cases.

MbedProgramData: holds metadata files associated with an Mbed program.

MbedLibReference: holds metadata associated with an Mbed library dependency.

### Test Coverage
- [x]  This change is covered by existing or additional automated tests.
- [ ]  Manual testing has been performed (and evidence provided) as automated testing was not feasible.
- [ ]  Additional tests are not required for this change (e.g. documentation update).

### CODE REVIEW

File: docs/index.html [hunk 1]
```diff
@@ -1,3 +1,7 @@
+<!--
```
Comment (rwalton-arm) [inline docs/index.html:1]: Added this to shut the license checker up.

File: news/20200330.feature [hunk 2]
```diff
@@ -0,0 +1 @@
+Add MbedProject classes
```
Comment (rwalton-arm) [inline news/20200330.feature:1]:
```suggestion
Add MbedProgram and metadata classes
```

**Comment by rwalton-arm** [comment]
> Looks awesome!
> 
> Would it be possible to expose `MbedProgram` as public api in `__init__.py` like we usually do?
> What about `exceptions`?

Oh yeah, I'll expose `MbedProgram` in the `__init__.py`.  I'll also add the `exceptions`. I was planning on doing that later to reduce PR size, but it makes sense to just add them now.

File: mbed_project/_internal/program.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+from mbed_project._internal.project_data import MbedProgramData
+
+logger = logging.getLogger(__name__)
+
+
+class MbedProgram:
+    """Represents an Mbed program.
+
+    An `MbedProgram` consists of a git repo and `MbedProgramData`. The `MbedProgramData` contains configuration files
```
Comment (rosiemiller2) [inline mbed_project/_internal/program.py:20]: This is really nice 👍

**Pull request merged**